### PR TITLE
Support tab movement/reordering

### PR DESCRIPTION
**Pull request by ssk97**
UI is to long-touch, then once selected, start moving.
~UI is to long touch on a tab already selected. This will start dragging all tabs that are selected. Grouped tabs can't be dragged for now.~

startDragAndDrop requires API 24 and the older function startDrag is deprecated. I don't know what your policy on that is. Right now I'm just using the deprecated one and suppressing the warning.

Potential improvements:
1) ~The drag shadow right now is only the tab you're long-touching, even if there are multiple tabs selected.~ Solved by removing multi-tab movement
2) ~There is no feedback during the drag/drop operation, perhaps add a highlight and/or a marker to indicate where the items will be dropped.~ Feedback added
3) Private tabs do not support selection and as such can't be dragged. They should probably start dragging immediately?
4) ~Deal with grouped tabs better. Right now I mostly block moving them so it shouldn't cause any problems, but I don't really understand the intent of the grouped tabs feature so I'll wait for direction there.~ Disable entirely if grouped tabs are enabled

Tests were added in the AC changes, mozilla-mobile/android-components#10936.

screenshot_dragging

This will cover most of #10344 (but with the new tab grouping, there may be further work needed to fully close that issue)

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Review by csadilek — COMMENTED** [review]
Thank you, @ssk97. This works really well! As a first step, let's get your A-C PR landed, see review there.

I've reached out to our product and UX teams to discuss and review this functionality. Please stay tuned as we discuss internally.

**Comment by ssk97** [comment]
I need help dealing with the build system/API deprecation. startDragAndDrop requires API 24, but the project has a minimum of API 21. Adding a version check to use the deprecated API causes the build to fail due to using a deprecated function. I'm unfamiliar with your processes so I'm unsure as to how to deal with this issue.

**Comment by eUgEntOptIc44** [comment]
@ssk97 any news on this?

**Comment by topotropic** [comment]
Sorry for the delay on this!

This is great work, thanks so much for picking it up. 

I filed two follow-up issues from a UX point of view:
https://github.com/mozilla-mobile/fenix/issues/22265
https://github.com/mozilla-mobile/fenix/issues/22264

Let me know if you have any questions! 

We still need to look into how the tab reordering could work with search groups that you find in Nightly – TBD.

**Comment by ssk97** [comment]
> Let me know if you have any questions!
> 
> We still need to look into how the tab reordering could work with search groups that you find in Nightly – TBD.

1- "Make Space" I tested it out, but there are weird animation/user-feel issues when moving vertically in grid mode and at the grid's horizontal boundaries. I'm sure they could be worked out, but I don't have the UX background necessary to easily figure out exactly how it should work (I think I have to adjust the target position based on where the tabs are coming from, which could be a problem with multi-tab dragging). In list mode it works just fine. Copy the ACTION_DROP code in AbstractBrowserTrayList to also run during ACTION_DRAG_LOCATION to see what happens.

2- Long touch/drag combo. I couldn't figure out a clean/easy way to do it. As far as I can tell I'd have to not use the built-in listeners and instead deal with the touch events directly. I'll work on figuring out how to do this.

Dealing with tab groups is definitely a major issue, but now that there is a disable switch for it, perhaps I should add a check that only allows dragging at all if tab groups are disabled? I currently block dragging the grouped tabs but I've found some other weird interactions with the feature, so probably safer to block them overall for now.

We'll want to support it at some point in the future once tab groups are more stabilized but I'd prefer to get a minimum viable version of this feature in first so that I can finally stop doing the incredibly annoying workarounds I currently need to do.

**Comment by topotropic** [comment]
Thanks for looking into all of that and the quick response!

1) "Make space"
> which could be a problem with multi-tab dragging

I'm not sure we need the capability to drag multiple tabs at a time – we could ignore it for now if that would make things easier. I made a quick screen recording of how old Fennec used to do it, something like this would be neat. 

https://user-images.githubusercontent.com/353960/140335704-1b27c387-2d75-4e1d-9f1c-cbf4db254b8f.mp4


2) Long touch/drag combo. It sounds that ignoring multi-select would make things easier so maybe that's a good idea to not worry about it now. 

> My understanding is that the UI will be: "wait for a long-touch event to happen on an unselected tab, selecting that tab. After that, if the user moves their finger sufficiently far before lifting, deselect all tabs and start the drag/drop operation". Is this correct?

Yes, I think so. Sebastian mentioned in the [other ticket ](https://github.com/mozilla-mobile/fenix/issues/22264#issuecomment-959321159) that Google Keep has a very similar interaction, check it out here to get a better idea of how this could work https://play.google.com/store/apps/details?id=com.google.android.keep&hl=en&gl=US 

> Dealing with tab groups is definitely a major issue, but now that there is a disable switch for it, perhaps I should add a check that only allows dragging at all if tab groups are disabled?

Yes, sounds good to only allow dragging if tabs groups are disabled.

Thank you!

**Comment by ssk97** [comment]
Ok, I've mostly completed both of those requests. The UI long-click system seems complete, though I haven't tested it extensively. The "make space" system still has a couple issues, and I'll have to fully rework my approach there.

I'm currently hiding the original tab (actually 0.2 alpha so not quite fully but that's for debug purposes, just set to 0.0 instead to make it fully invisible) and creating a drag shadow. This works well enough usually, but has edge case issues. If dragged outside of the tab tray, the drag shadow will automatically animate a failed drag despite the tab being moved anyway. If a drag is dropped right after a movement, the invisible swap position animation is made visible and it looks noticeably wrong. If the tab updates after the drag is started, it seems to override the invisibility.

I think these are all very minor issues, but I don't see any reasonable way to resolve them all with this design. As such, I think the correct solution is to instead apply a translation to the tab's ViewHolder, updating it constantly while the drag is going. If done correctly, this shouldn't have any of the problems of the drag shadow+invisibility method. However, this is a much more invasive change. To do this I think would involve creating an overridden LayoutManager for the AbstractBrowserTrayList to use, but I'll need to spend more time investigating to determine for certain.

**Comment by ssk97** [comment]
Ok, drag "make space" behavior complete and I fixed a bug in the long-click then drag behavior. I made it so that dragging can't occur if multiple tabs are selected. The only remaining issue is that if you have enough tabs open to cause scrolling, there's some very strange behavior that will need to be fixed.

The easiest to fix should be the scroll caused by the initial long-click. I think I found what's causing that and I assume it won't be a problem to disable the scrolling for the first item added to the selection group. The larger problem is that sometimes dragging down seems to cause the tab tray to quickly scroll all the way to the very bottom, not sure what's causing that.

I'll also need to add some intended scrolling behavior so that the tray scrolls when the user drags a tab near the edge.

**Comment by topotropic** [comment]
Thanks! I briefly tried it out and it's looking really good. 

I notices two things: scrolling the tab tray when you drag a tab isn't working – I think you mentioned that also and multi-select stops working after I dragged a tab, so I can't multi-select on long pressing.

Ping me if you need a review or feedback.

**Comment by ssk97** [comment]
Right now I've disabled the scrolling on tab selection entirely.  It was done with a scroll on bind() in AbstractBrowserPageViewHolder, but it only seemed to work half the time anyway. If you'd like that behavior re-introduced I can do so, I'd put it with the selection logic instead of hidden away in the bind() call.
I may have misunderstood the purpose of that scroll call, if so please let me know what the intended behavior there was.

I noticed that some strange behavior can occur if a tab updates (such as from finishing loading) while you're on the tab tray. In my area I'm constantly setting the beingDragged and elevation values so that it'll just be a momentary flicker instead of a larger issue, but there seems to be some issues with the selection logic if that happens.

@topotropic 
I'm ready for more feedback/review.

> multi-select stops working after I dragged a tab, so I can't multi-select on long pressing.
Can you give more details or a video demonstrating this? I tried and long press into multi-select seems to work the same before and after I've done a drag.

**Comment by topotropic** [comment]
Thanks! It's already looking pretty good. 

I made a screenrecording with the multi-select issue – so I can mutli-select the first time but when I try it again and longpress on a card, it just briefly flickers but doesn't go into multi-select mode. 

https://user-images.githubusercontent.com/353960/143073831-756e76bb-1c43-4328-b017-9659d2918678.mp4

**Comment by ssk97** [comment]
Thanks for the video. It looks like the problem was that the drag event was starting a frame after the selection, which makes me think the problem was your finger moving a pixel or two. I've added in some custom logic to disallow dragging until you've moved far enough, which should fix the problem.

**Comment by ssk97** [comment]
I just found a reliable reproduction of the scrolling bug. On an emulated Nexus 7 with default display size with grid view, if you scroll to the very bottom and then long-click a tab, the tray scrolls up 1 row. Further interactions will also cause additional scroll, it doesn't matter which tab you're selecting or deselecting. This will continue until either you manually scroll (misaligning it) or it hits the top of the tab tray.

This happens in the current main branch, so it's not my code causing it. I thought that I had fixed it by removing the scroll in the bind() call, but it seems like it's still happening in some circumstances. As such I'm reverting that change.

Seeing as this only causes a minor inconvenience and seems to happen in very specific circumstances, and the fact that it's a bug in pre-existing code, I don't think this is something I'll be fixing.

**Review by pocmo — COMMENTED** [review]
This works very well on my device. Very nice. Thank you for this! 👍 

I only added some nits/questions. Overall I think that looks great.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt
Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt:252]: It shouldn't be null but there is a possibility that the parent is something other than AbstractBrowserTrayList (Not in the current code AFAIK, but someone could try to do code reuse). I think crashing makes sense in that scenario, so I'll change it to an unsafe cast.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt
Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt:256]: Does checking settings take long? I could do the check on ACTION_DOWN I suppose, but I think it makes more logical sense here and touchStartPoint isn't obviously related to search tab groups.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt
Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt:273]: Thanks! Will fix.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt
Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt:266]: 30 was chosen via guess-and-check. It just had to be large enough to not be triggered by the touchscreen jittering the touch point and triggering ACTION_MOVE, which seemed to be causing problems for topotropic.

I'm more worried about the scroll speed and area in AbstractBrowserTrayList, as those are definitely affected by device sizes. The numbers I chose there seem to work on the devices I've tried/emulated, but a more elegant solution would be good.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt
Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt:273]: Yeah, I don't see an easy way to fix that.

Comment (pocmo) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt]: Just found `ViewConfiguration.getScaledTouchSlop()`: https://developer.android.com/reference/android/view/ViewConfiguration#getScaledTouchSlop() - maybe that could help here?

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt:267]: Just found `ViewConfiguration.getScaledTouchSlop()`: https://developer.android.com/reference/android/view/ViewConfiguration#getScaledTouchSlop() - maybe that could help here?

Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt]: That works, good find. I've made the change.

**Review by pocmo — APPROVED** [review]
LGTM.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt:258]: Thank you @ssk97 for moving this forward and bearing with us! We would like to enable this feature in Nightly only for now to gather more feedback. Would you mind adding a `tabReorderingFeature` flag to `FeatureFlags.kt, set to `Config.channel.isNightlyOrDebug`, and checking if true here as well?

Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt]: Done. I had to separate out the function since the added condition hit the complexity limit of detekt.

Comment (csadilek) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTabViewHolder.kt]: Thanks, @jonalmeida offered to take another look, and pending his review, we hope to land it in Nightly next week.

File: app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayInteractor.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayInteractor.kt:41]: nit: this would be nice to have on the `BrowserTrayInteractor`. The TabsTrayInteractor is generally for "global" tray interactions that apply that would also apply to Synced Tabs as well.

**Review by jonalmeida — APPROVED** [review]
Looks great! ✨ 

I only have one concern about detaching our listeners safely, but I think we can land this otherwise.

File: app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayInteractor.kt
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayInteractor.kt:41]: It might not be worth moving it at this point though - just a note.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTrayList.kt
Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTrayList.kt:45]: That's what `this.setOnDragListener(null)` in `onDetachedFromWindow` is meant to do. Am I misunderstanding something?

File: app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayInteractor.kt
Comment (ssk97) [reply app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayInteractor.kt:41]: The issue with that is the `AbstractBrowserTrayList` doesn't seem to have a `BrowserTrayInteractor`. I could add one but that didn't seem reasonable.

Also, shouldn't a tab reordering be represented in synced tabs? I would expect the tabs list to remain in the same order between devices, and if rearranged the movement would be copied to the other on the next sync.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTrayList.kt
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTrayList.kt:45]: Ah, sorry I didn't see that. Looks good to me then!

**Review by csadilek — APPROVED** [review]
Thanks again, @ssk97.

We have some build/tooling limitations so needed to open a new PR to land this. Still your commits and you're the author, but the PR is here: https://github.com/mozilla-mobile/fenix/pull/22751

Once green it will land and we can close this one.

**Comment by pocmo** [comment]
It's in. 🛬

https://github.com/mozilla-mobile/fenix/commit/6637f610884bf4c6826886b4a413a23f3400622e

**Comment by marcel263** [comment]
With respect for developer it should be considered: what's the point of this if it does not work with the defaukts Firefox ships with (grouped tabs)? There will be little exposure. Adding tabs to a group by dragging could be a beneficial feature.

**Comment by doodhout** [comment]
Just to keep people up-to-date and not have everyone seperately digging in commit lists and release tags: this seems to not be included in the latest Fenix v96 beta, in other words it will probably be included in v97 at the earliest.

The current android nightly (v97a1) does not seem to have the functionality, as far as I can tell. I don't think nightlies are feature-frozen yet, so perhaps soon?

**Comment by ssk97** [comment]
> The current android nightly (v97a1) does not seem to have the functionality, as far as I can tell. I don't think nightlies are feature-frozen yet, so perhaps soon?

The current nightly does, I've been using it quite happily. However, it only works if you disable tab groups (nightly has tab groups enabled by default). On beta and release, there's a feature flag which disables this feature at all times, and I have no info on if/when it will be enabled outside of nightly.

If you're having any further problems with it, please let me know!

**Comment by doodhout** [comment]
> The current nightly does, I've been using it quite happily. However, it only works if you disable tab groups (nightly has tab groups enabled by default).

You are right, I had not realised. I confirm that tab reordering works in the nightly when you first disable tab grouping in the tab settings (i.e. not the regular settings, but while in the tab overview you go into the three dots and there into the tab settings).

Nice, I can't wait for this feature to land since that's one of the last deal breakers for me (this and keyword search).

**Comment by doodhout** [comment]
Fenix beta 97 is just released. Unfortunately tab reordering doesn't seem to be enabled on beta. Tab groups are not active (just like you said). Maybe the feature must first be activated (via feature flag) and then disabled (via user setting) before tab reordering will work in beta/stable? Or is there another plan to introduce tab reordering in beta/stable?

**Comment by adrianinsaval** [comment]
what happened with this feature? I was using brave browser but now it also switched to a useless list view of tabs without reordering or grouping support so I wanted to move back to firefox, but now I find firefox doesn't seem to have this feature either, there is no visible option to group tabs either. Can anybody explain the status of this please?

**Pull request closed without merging**
